### PR TITLE
Effects to rust

### PR DESCRIPTION
**Pull request by chriseth**
Formats a vector of "effects" coming from the witgen solver into rust code, compiles and loads it.

Submachine calls and receiving arguments will be done in another PR.

This code assumes `known` to be a padded bit vector ( #2230 ).

### CODE REVIEW

File: executor/src/witgen/jit/compiler.rs [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+}
+
+pub type WitgenFunction<T> = extern "C" fn(WitgenFunctionParams<T>);
+
+pub fn compile_effects<T: FieldElement>(
+    first_column_id: u64,
+    column_count: usize,
+    known_inputs: &[Cell],
+    effects: &[Effect<T, Cell>],
+) -> Result<(Arc<Library>, WitgenFunction<T>), String> {
```
Comment (chriseth) [inline executor/src/witgen/jit/compiler.rs:32]: We need to keep the `Library` in an Arc because otherwise it will be unloaded. We can change the interface once we know where and how the loaded functions will be stored.

File: executor/src/witgen/jit/compiler.rs [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+        let effects = vec![
+            assignment(&x0, number(7) * symbol(&a0)),
+            assignment(&ym1, symbol(&x0)),
+            assignment(&yp1, symbol(&ym1) + symbol(&x0)),
+            Effect::Assertion(Assertion {
+                lhs: symbol(&ym1),
+                rhs: symbol(&x0),
+                expected_equal: true,
+            }),
+        ];
```
Comment (georgwiese) [inline executor/src/witgen/jit/compiler.rs:424]: Should we have a slightly more complex example to be used in both tests? It would be nice to test each operation at least once as well as edge cases like field division and negative field elements.

File: executor/src/witgen/jit/compiler.rs [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+    symbolic_expression::{BinaryOperator, SymbolicExpression, UnaryOperator},
+};
+
+#[repr(C)]
+pub struct WitgenFunctionParams<T> {
+    data: *mut T,
+    known: *mut u32,
+    len: u64,
+    row_offset: u64,
+    call_machine: extern "C" fn(u64, *mut LookupCell<'_, T>, usize) -> bool,
```
Comment (georgwiese) [inline executor/src/witgen/jit/compiler.rs:22]: Please at docstrings here.

File: executor/src/witgen/jit/compiler.rs (hunk 1 above)
Comment (georgwiese) [reply executor/src/witgen/jit/compiler.rs:32]: Do I understand this right, the returned library needs to stay in the scope as long as the function might be used? If yes, please add this to the docstring :)

File: executor/src/witgen/jit/compiler.rs [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+    let column_name = column_name
+        .rfind("::")
+        .map_or(column_name.as_str(), |i| &column_name[i + 2..])
+        .replace("[", "_")
+        .replace("]", "_");
+    if *row_offset < 0 {
+        format!("{column_name}_u{}", -row_offset)
+    } else {
+        format!("{column_name}_d{row_offset}")
+    }
```
Comment (georgwiese) [inline executor/src/witgen/jit/compiler.rs:195]: Why u and d?

File: executor/src/witgen/jit/compiler.rs [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    fn sub(self, b: Self) -> Self {{
+        // TODO this is inefficient.
+        Self(IntType::try_from((DoubleIntType::from(self.0) + DoubleIntType::from(MODULUS) - DoubleIntType::from(b.0)) % DoubleIntType::from(MODULUS)).unwrap())
+    }}
+}}
+impl std::ops::Mul<FieldElement> for FieldElement {{
+    type Output = Self;
+    #[inline]
+    fn mul(self, b: FieldElement) -> FieldElement {{
+        Self(IntType::try_from((DoubleIntType::from(self.0) * b.0 as DoubleIntType) % DoubleIntType::from(MODULUS)).unwrap())
```
Comment (georgwiese) [inline executor/src/witgen/jit/compiler.rs:264]:
```suggestion
        Self(IntType::try_from((DoubleIntType::from(self.0) * DoubleIntType::from(b.0)) % DoubleIntType::from(MODULUS)).unwrap())
```
Is there a difference between the two?

File: executor/src/witgen/jit/compiler.rs [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+        "u64"
+    };
+    let modulus = T::modulus();
+
+    Ok(format!(
+        r#"#![allow(non_snake_case)]
+
+#[derive(Clone, Copy, Default)]
+#[repr(transparent)]
+struct FieldElement({int_type});
```
Comment (georgwiese) [inline executor/src/witgen/jit/compiler.rs:227]: Is there a plan to use optimized field implementations, like we have for Goldilocks?

**Review by georgwiese — COMMENTED** [review]
Nice!

File: executor/src/witgen/jit/compiler.rs (hunk 2 above)
Comment (chriseth) [reply executor/src/witgen/jit/compiler.rs:432]: Not sure I really want to test every operator, but probably a good idea...

File: executor/src/witgen/jit/compiler.rs (hunk 4 above)
Comment (chriseth) [reply executor/src/witgen/jit/compiler.rs:195]: `up` and `down`. Please suggest a better name :)

File: executor/src/witgen/jit/compiler.rs (hunk 5 above)
Comment (chriseth) [reply executor/src/witgen/jit/compiler.rs:265]: Haha interesting. I don't think so, those are primitive types.

File: executor/src/witgen/jit/compiler.rs (hunk 6 above)
Comment (chriseth) [reply executor/src/witgen/jit/compiler.rs:227]: Yes, but I think we have to re-implement it here.

File: executor/src/witgen/jit/compiler.rs [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+    let y_d1 = (y_u1 + x_d0);
+    assert!(y_u1 == x_d0);
+
+    set(data, row_offset, 0, 0, x_d0);
+    set(data, row_offset, -1, 1, y_u1);
+    set(data, row_offset, 1, 1, y_d1);
+
+    set_known(known, row_offset, 0, 0);
+    set_known(known, row_offset, -1, 1);
+    set_known(known, row_offset, 1, 1);
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/jit/compiler.rs:464]: The main reason I did it like this is because I think this is the best form for the optimizer: First copy everything to local variables (i.e. the stack), operate only on stack variables and then copy to memory again. My assumption is that as long as stuff is on the stack, the optimizer knows that it cannot be modified. If data is in memory, any memory write can potentially destroy the data and we cannot re-use already computed values. This is particularly important for submachine calls: If values are on the stack, we can be sure that the are still the same after the call to the submachine. Also I think the optimizer is smart enough to directly store data in memory if it will not be touched again.

We should verify all these assumptions by looking at the assembly an the performance, of course.

File: executor/src/witgen/jit/compiler.rs (hunk 4 above)
Comment (georgwiese) [reply executor/src/witgen/jit/compiler.rs:195]: Don't have a great idea. Maybe I'd just prepend `n` for "negative", and nothing for the offsets >= 0 (the typical case).

File: executor/src/witgen/jit/compiler.rs (hunk 1 above)
Comment (chriseth) [reply executor/src/witgen/jit/compiler.rs]: Fixed

File: executor/src/witgen/jit/compiler.rs [hunk 8]
```diff
@@ -0,0 +1,804 @@
...
+        let effects = vec![
+            assignment(&cell("x", 0, 0), number(8).field_div(&number(2))),
+            assignment(&cell("x", 0, 1), number(1 << 60).field_div(&number(8))),
+            assignment(&cell("x", 0, 2), (-number(8)).field_div(&number(2))),
+            assignment(&cell("x", 0, 3), number(8).field_div(&-number(2))),
+            assignment(&cell("x", 0, 4), (-number(8)).field_div(&-number(2))),
+        ];
+        let f = compile_effects(0, 1, &[], &effects).unwrap();
+        let mut data = vec![GoldilocksField::from(0); 5];
+        let mut known = vec![0; 1];
```
Comment (georgwiese) [inline executor/src/witgen/jit/compiler.rs:709]:
```suggestion
        let mut known = vec![0; 5];
```
We need at least one `u32` per row. This fixes a segmentation fault when I run on my Mac.

**Review by georgwiese — APPROVED** [review]
Cool! Just some small comments.

File: jit-compiler/src/compiler.rs [hunk 9]
```diff
@@ -224,6 +231,15 @@ pub fn call_cargo(code: &str) -> Result<PathInTempDir, String> {
...
         }
     }
+    #[allow(clippy::print_stdout)]
+    if output_asm {
+        let asm_file = dir
+            .join("target")
+            .join("release")
+            .join("deps")
+            .join("powdr_jit_compiled.s");
+        println!("{}", fs::read_to_string(&asm_file).unwrap());
```
[the remark this answers is not in the document]
Comment (chriseth) [reply jit-compiler/src/compiler.rs:241]: I was not able to get any logging output from that crate - at least when running tests.

**Review by georgwiese — APPROVED** [review]
